tpm2/src/util/nsl_util.c

```diff
@@ -0,0 +1,983 @@
+    *nonce = NULL;
+    *nonce_len = 0;
+
+    *id_len = 0;
+
+    memset(message, 0, message_len);
+    free(message);
+
+    return 1;
+  }
+  memcpy(*id, index, *id_len);
+
+  memset(message, 0, message_len);
+  free(message);
+
+  return 0;
+}
+
+//
+// build_nonce_response()
+//
+int build_nonce_response(EVP_PKEY_CTX * ctx,
+                         unsigned char *nonce_a, size_t nonce_a_len,
+                         unsigned char *nonce_b, size_t nonce_b_len,
+                         unsigned char *id, size_t id_len,
+                         unsigned char **response, size_t *response_len)
+{
+  // Allocate the unencrypted response buffer.
+  unsigned char *message = NULL;
+  size_t message_len =
+    id_len + nonce_a_len + nonce_b_len + (3 * sizeof(size_t));
+  message = calloc(message_len, sizeof(unsigned char));
+  if (message == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the message buffer.");
+    return 1;
+  }
+
+  // Build the nonce response message.
+  unsigned char *index = message;
+
+  memcpy(index, &nonce_a_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, nonce_a, nonce_a_len);
+  index += nonce_a_len;
+  memcpy(index, &nonce_b_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, nonce_b, nonce_b_len);
+  index += nonce_b_len;
+  memcpy(index, &id_len, sizeof(size_t));
+  index += sizeof(size_t);
+  memcpy(index, id, id_len);
+
+  // Encrypt the nonce response and then clean up the unencrypted response.
+  int result =
+    encrypt_with_key_pair(ctx, message, message_len, response, response_len);
+
+  memset(message, 0, message_len);
+  free(message);
+
+  // Handle encryption errors if any occurred.
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to encrypt the nonce response.");
+    return 1;
+  }
+
+  return 0;
+}
+
+//
+// parse_nonce_response()
+//
+int parse_nonce_response(EVP_PKEY_CTX * ctx,
+                         unsigned char *response, size_t response_len,
+                         unsigned char **nonce_a, size_t *nonce_a_len,
+                         unsigned char **nonce_b, size_t *nonce_b_len,
+                         unsigned char **id, size_t *id_len)
+{
+  // Decrypt the nonce response.
+  unsigned char *message = NULL;
+  size_t message_len = 0;
+  int result =
+    decrypt_with_key_pair(ctx, response, response_len, &message, &message_len);
+
+  if (result)
+  {
+    kmyth_log(LOG_ERR, "Failed to decrypt the nonce response.");
+    return 1;
+  }
+  unsigned char *index = message;
+
+  // Parse out the first nonce.
+  memcpy(nonce_a_len, index, sizeof(size_t));
+  index += sizeof(size_t);
+  *nonce_a = calloc(*nonce_a_len, sizeof(unsigned char));
+  if (*nonce_a == NULL)
+  {
+    kmyth_log(LOG_ERR, "Failed to allocate the first nonce buffer.");
+
+    *nonce_a_len = 0;
+
+    memset(message, 0, message_len);
+    free(message);
+
+    return 1;
+  }
+  memcpy(*nonce_a, index, *nonce_a_len);
+  index += *nonce_a_len;
+
+  // Parse out the second nonce.
```

Comment:
Should we validate the length of received nonce_b to avoid parsing an invalid value?

Reply:
Ditto